Makefile

```
@@ -64,27 +67,88 @@ build-env:
+
+	cd nx-X11 && make BuildDepends FONT_DEFINES=$(FONT_DEFINES)
+
+clean-env: version
+	${RM_FILE} nx-X11/exports/include/nx-X11/Xlib.h			\
+	      nx-X11/exports/include/nx-X11/Xresource.h			\
+	      nx-X11/exports/include/nx-X11/Xutil.h			\
+	      nx-X11/exports/include/nx-X11/cursorfont.h		\
+	      nx-X11/exports/include/nx-X11/Xlibint.h			\
+	      nx-X11/exports/include/nx-X11/Xcms.h			\
+	      nx-X11/exports/include/nx-X11/Xlocale.h			\
+	      nx-X11/exports/include/nx-X11/XKBlib.h			\
+	      nx-X11/exports/include/nx-X11/XlibConf.h			\
+	      nx-X11/exports/include/nx-X11/Xregion.h			\
+	      nx-X11/exports/include/nx-X11/ImUtil.h			\
+	      $(NULL)
+	${RM_FILE} nx-X11/exports/include/nx-X11/Xtrans/transport.c	\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtrans.c		\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtrans.h		\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtransint.h		\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtranslcl.c		\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtranssock.c		\
+	      nx-X11/exports/include/nx-X11/Xtrans/Xtransutil.c		\
+	      $(NULL)
+	-rmdir exports/include/nx-X11/Xtrans/
+	-rmdir exports/include/nx-X11/
+
 	# prepare Makefiles and the nx-X11 symlinking magic
-	cd nx-X11 && make BuildEnv FONT_DEFINES=$(FONT_DEFINES)
+	cd nx-X11 && make CleanEnv FONT_DEFINES=$(FONT_DEFINES)
 
 build-lite:
 	cd nxcomp && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE}
 	cd nxproxy && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE}
 
 build-full: build-env
 # in the full case, we rely on "magic" in the nx-X11 imake-based makefiles...
+
+	# build nxcomp first
 	cd nxcomp && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE}
-	# build libNX_X11 and libNX_Xext prior to building
-	# nxcomp{ext,shad}.
-	cd nx-X11/lib && make
 
-	cd nxcompshad && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE}
+	# build libNX_X11 second
+	cd nx-X11/lib && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE} && cd - || exit -1
```

We don't actually need `cd - || exit -1` here, since each line is executed in a subshell (so CWD changes won't reflect in upper layers) and the exit code of each line is also checked by make. If any fail, the whole target fails.

Was there any specific reason why you did this, or just precaution?

---

> We don't actually need cd - || exit -1 here

Yes, we do. I want the overall make process to fail, if any of the sub-makes fails. 

If I do ``${MAKE} && cd -1 || exit 1`` it will result in either cd - on build success or on exit, if builds fail.

---

Yes, but that'll happen with a single `${MAKE}` call as well? Its return value is passed up, so a non-zero one makes the current target fail. Even if it succeeds, we don't need to cd, since each line in the recipe is run in a new shell process (and thus has the same initial CWD.)

Try to change this to `cd nx-X11/lib && autoreconf -vfsi && (${CONFIGURE}) && ${MAKE}` and make the build fail explicitly in libNX_X11. Should work as you expect it.

Then, remove the explicit fault again and add `	pwd` before the `mkdir` line to see that the initial CWD is unchanged for the `pwd` call.